utils/pattern/src/frontend/mod.rs

```diff
@@ -9,17 +9,57 @@ mod serde;
 use crate::common::*;
 use crate::Error;
 use crate::PatternOrUtf8Error;
+use crate::SinglePlaceholderPattern;
 #[cfg(feature = "alloc")]
 use crate::{Parser, ParserOptions};
 #[cfg(feature = "alloc")]
 use alloc::{borrow::ToOwned, str::FromStr, string::String};
+use core::ops::Deref;
 use core::{
     convert::Infallible,
     fmt::{self, Write},
     marker::PhantomData,
 };
 use writeable::{adapters::TryWriteableInfallibleAsWriteable, PartsWrite, TryWriteable, Writeable};
 
+#[allow(clippy::unwrap_used)]
+#[cfg(feature = "alloc")]
+impl SinglePlaceholderPattern<str> {
+    pub fn encode_store(prefix: &str, suffix: &str) -> alloc::boxed::Box<str> {
+        alloc::format!(
+            "{}{prefix}{suffix}",
+            char::from_u32(prefix.len() as u32 + 1).unwrap()
+        )
+        .into_boxed_str()
+    }
+}
+
+#[allow(clippy::unwrap_used, clippy::indexing_slicing)]
+impl<Store> SinglePlaceholderPattern<Store>
+where
+    Store: Deref<Target = str>,
```

Comment:
Issue: The bound I use everywhere else is `AsRef<str>`.

Reply:
I don't like that bound.

Reply:
We should open a discussion about whether we want the backend to be

1. `AsRef<str>`
2. `Deref<Target = str>`
3. `Borrow<str>`
4. `str`

But for now please remain consistent.

Reply:
5. `Backend::Store`

Reply:
Yes, for all of those I mean a substitution of `str` with `Backend::Store`

Reply:
Just to start the commentary a bit:

I went with `AsRef<Backend::Store>` because it is the most flexible.

Why AsRef: The `Store` parameter worked out well for me in LiteMap and ZeroTrie before embarking on Pattern. It makes certain impls a pain to write, but once they are written, things "just work" and are very easy to use and very flexible.

Why not Deref: I don't care that the store dereferences uniquely to `Backend::Store`. For instance, at some point we might get `Box<AsciiStr>` which derefs to `AsciiStr` before `str`.

Why not Borrow: I don't care about the additional invariants on Borrow like Hash and Eq. If I want to implement those traits on Pattern, I can; I don't need Borrow's help.

Why not DSTs: I didn't go with the dynamically sized type because I dislike the amount of `unsafe` code required for the DSTs to work. I have made multiple attempts to write safe abstractions for ref casts and box casts, but they keep being shot down (#5101). I would like to block the pure DST approach on first solving #2778 / #2310 / #5127, but I'm open to it assuming it can be written in a way that minimizes `unsafe`.